Add streamline Logger

DavidLegg opened the pull request:
* **Tickets addressed:** N/A
* **Review:** By commit  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description
Adds a Logger class to the streamline framework, which uses topics to emit log messages directly as events. Also builds such a Logger when the Registrar is initialized, and reconfigures errors to go to that logger when the error behavior is set to LOG. Before this, errors were collected awkwardly into a discrete string resource.

## Verification
<!-- How were the changes validated? Were any automated tests added, updated, removed, or re-baselined? -->
Since the logger is connected to the Registrar's error behavior, this was tested using the streamline-demo example model. By adding "CauseError" activities to the plan, we can trip the resource error handling machinery, and verify that appropriate error messages appear in the Simulation Events tab.

<img width="1792" alt="image" src="https://github.com/NASA-AMMOS/aerie/assets/13932065/a97a9009-521e-4738-9960-d5627c5a58e1">

## Documentation
<!-- What documentation was invalidated by these changes? Which artifacts should reviewers check for accuracy and completeness? -->
None yet - #1494 should probably be updated to tell users how to use the logger in their model/activities.

This does remove the `errors` resource, but it leaves the `numberOfErrors` resource intact, which we should note for any users who may have depended on `errors`.

## Future work
<!-- What next steps can we anticipate from here, if any? -->
None planned

Commit:
Add streamline Logger

Adds a Logger class to the streamline framework, which uses topics to emit log messages directly as events.
Also builds such a Logger when the Registrar is initialized, and reconfigures errors to go to that logger when the error behavior is set to LOG.
Before this, errors were collected awkwardly into a discrete string resource.

Commit:
Move primary logger to Logging

Move the primary Logger instance to a separate class, out of Registrar.
This makes more sense logically, as Logging only incidentally depends on the Registrar, in the sense that creating the Registrar is a natural place to initialize all of the singletons.
It also makes the logger easier to find (I hope).

mattdailis committed:
Use UNIT to implement no-op effects

mattdailis commented on contrib/src/main/java/gov/nasa/jpl/aerie/contrib/streamline/debugging/Logger.java:
```diff
@@ -0,0 +1,60 @@
+    private final Instant planStart;
+
+    public Logger(Registrar registrar, Instant planStart) {
+        // TODO - refactor the plan start out, by instead building a global sim clock to do the same job.
```
@DavidLegg and I discussed that this TODO comment should be addressed before merging this PR. In particular - there's no need to explicitly include the log timestamp in the message, because every serializable event is time-tagged

mattdailis reviewed (changes requested):
See https://github.com/NASA-AMMOS/aerie/pull/1500/files#r1705853363

mattdailis commented on contrib/src/main/java/gov/nasa/jpl/aerie/contrib/streamline/modeling/Registrar.java:
```diff
@@ -64,17 +59,13 @@ public enum ErrorBehavior {
   }
 
-  public Registrar(final gov.nasa.jpl.aerie.merlin.framework.Registrar baseRegistrar, final ErrorBehavior errorBehavior) {
+  public Registrar(final gov.nasa.jpl.aerie.merlin.framework.Registrar baseRegistrar, final Instant planStart, final ErrorBehavior errorBehavior) {
```
If we can avoid changing the signature of the Registrar constructor, I think this PR can be backwards compatible with code that already uses the streamline framework

Commit:
Remove time from logging

Per @mattdailis' suggestion, since serialized events are tagged with the time they're emitted already, there's no need to include that information in the log message itself.
Further, removing that dependency on the plan start time makes this PR fully backwards-compatible.

mattdailis commented on contrib/src/main/java/gov/nasa/jpl/aerie/contrib/streamline/modeling/Registrar.java:
```diff
@@ -66,27 +56,11 @@ public enum ErrorBehavior {
-    return ExceptionUtils.stream(e)
-        .map(ExceptionUtils::getMessage)
-        .collect(joining("\nCaused by: "));
+    discrete("numberOfErrors", numberOfErrors, new IntegerValueMapper());
```
In the event that users have written constraints that depend on the previous resource definition, it may be worth calling out this change in the release notes. Not a reason not to make this change, but worth communicating explicitly.

It also highlights that constraints (and scheduling?) would benefit from being able to refer to simulation events.

mattdailis reviewed (approved):
Thanks for making this backwards compatible! Looks good to me

EDIT: Resolving conversation below so it doesn't block merge, but copying up here for visibility:
> In the event that users have written constraints that depend on the previous resource definition, it may be worth calling out in the release notes [that the errors resources has been replaced with a numberOfErrors resource]. Not a reason not to make this change, but worth communicating explicitly.

DavidLegg commented:
One slight correction to that note on the `errors` and `numberOfErrors` resources - We actually had `numberOfErrors` already, just derived in a different way. This PR does remove `errors` itself, though, so I agree we should put that in the release notes just in case.
